File: perception-core/src/main/java/com/hh/uiperception/core/semantic/RefAssigner.java
package com.hh.uiperception.core.semantic;

/**
 * Assigns snapshot-local ref identifiers to interactive/executable nodes in a semantic tree.
 *
 * Ref strategy differs between web and native pipelines:
 * - Web mode: every node with valid bounds gets a ref (prefix "w1", "w2", ...)
 *   — the JS side already assigns __pr_idx for operation targeting
 * - Native mode: only nodes with interactive roles (button, input, checkbox, etc.)
 *   get refs — avoids polluting the snapshot with refs on non-interactive elements like text/image
 *
 * Special handling for LIST_ITEM in native mode (three-tier clickability logic):
 * 1. Directly clickable or inferred → always assign ref
 * 2. Clickability "guessed" (heuristic) → assign ref only if no executable descendants exist
 *    (avoids double-assigning when a child button already has a ref)
 * 3. Not clickable → no ref
 */
public final class RefAssigner {

    private RefAssigner() {
    }

    /** Assign refs with default prefix "n" in native mode. */
    public static SemanticNode assign(SemanticNode root) {
        return assign(root, "n", false);
    }

    /** Assign refs with custom prefix in native mode. */
    public static SemanticNode assign(SemanticNode root, String prefix) {
        return assign(root, prefix, false);
    }

    /**
     * Assign refs to the tree.
     * @param prefix   ref prefix ("n" for native, "w" for web)
     * @param webMode  true → all nodes with valid bounds get refs; false → only interactive nodes
     */
    public static SemanticNode assign(SemanticNode root, String prefix, boolean webMode) {
        if (root == null) {
            return null;
        }
        return new State(prefix, webMode).assignNode(root);
    }

    /**
     * Decide whether a node should get a ref.
     * Web: any node with valid bounds.
     * Native: role-based check, with special list-item logic.
     */
    private static boolean shouldAssignRef(SemanticNode node, boolean webMode) {
        if (node.bounds() == null || !node.bounds().isValid()) {
            return false;
        }
        if (webMode) {
            return true;
        }
        if (node.role() == SemanticRole.LIST_ITEM) {
            return shouldAssignRefToListItem(node);
        }
        return isExecutableRole(node.role());
    }

    /** 12 interactive roles that always get refs in native mode. */
    private static boolean isExecutableRole(SemanticRole role) {
        return role == SemanticRole.BUTTON
                || role == SemanticRole.INPUT
                || role == SemanticRole.CHECKBOX
                || role == SemanticRole.RADIO
                || role == SemanticRole.SWITCH
                || role == SemanticRole.SLIDER
                || role == SemanticRole.PICKER
                || role == SemanticRole.LIST
                || role == SemanticRole.GRID
                || role == SemanticRole.SCROLL
                || role == SemanticRole.WEBVIEW;
    }

    /**
     * List-item ref assignment: three-tier clickability logic.
     * - CLICKABLE / CLICKABLE_INFERRED → always assign
     * - CLICKABLE_GUESSED → assign only if no executable descendants (avoid double-assign)
     * - Not clickable → skip
     */
    private static boolean shouldAssignRefToListItem(SemanticNode node) {
        if (node.states().contains(SemanticStates.CLICKABLE)
                || node.states().contains(SemanticStates.CLICKABLE_INFERRED)) {
            return true;
        }
        if (node.states().contains(SemanticStates.CLICKABLE_GUESSED)) {
            return !hasExecutableDescendant(node);
        }
        return false;
    }

    /** Mutable state for recursive traversal: tracks the next ref number. */
    private static final class State {
        private final String prefix;
        private final boolean webMode;
        private int nextRef = 1;

        State(String prefix, boolean webMode) {
            this.prefix = prefix;
            this.webMode = webMode;
        }

        /** Recursively walk the tree, rebuilding each node with an assigned ref if eligible. */
        private SemanticNode assignNode(SemanticNode node) {
            SemanticNode.Builder builder = SemanticNode.builder(node.role())
                    .name(node.name())
                    .text(node.text())
                    .contentDescription(node.contentDescription())
                    .resourceId(node.resourceId())
                    .className(node.className())
                    .bounds(node.bounds())
                    .roleDecision(node.roleDecision())
                    .webElementIdx(node.webElementIdx());
            for (String state : node.states()) {
                builder.addState(state);
            }
            if (shouldAssignRef(node, webMode)) {
                builder.ref(prefix + nextRef++);
            } else if (node.hasRef()) {
                builder.ref(node.ref());
            }
            for (SemanticNode child : node.children()) {
                builder.addChild(assignNode(child));
            }
            return builder.build();
        }

    }

    /** Check if any descendant qualifies for a ref (used by list-item guess logic). */
    private static boolean hasExecutableDescendant(SemanticNode node) {
        for (SemanticNode child : node.children()) {
            if (shouldAssignRef(child, false) || hasExecutableDescendant(child)) {
                return true;
            }
        }
        return false;
    }
}

File: ios/UIPerception/Capture/RefAssigner.swift
import Foundation

/// 给语义树中的可交互节点分配 snapshot-local ref 标识。
/// 对应 Android `RefAssigner.java`：
/// - native mode：只有可交互角色（button/input/...）+ 三档 listitem 逻辑才分配
/// - web mode：所有有 bounds 的节点都分配
enum RefAssigner {
    static func assign(_ root: SemanticNode?, prefix: String = "n", webMode: Bool = false) -> SemanticNode? {
        guard let root else { return nil }
        return State(prefix: prefix, webMode: webMode).assignNode(root)
    }

    private static func shouldAssignRef(_ node: SemanticNode, webMode: Bool) -> Bool {
        guard let bounds = node.bounds, bounds.isValid else { return false }
        if webMode { return true }
        if node.role == .listItem {
            return shouldAssignRefToListItem(node)
        }
        return isExecutableRole(node.role)
    }

    private static func isExecutableRole(_ role: SemanticRole) -> Bool {
        switch role {
        case .button, .input, .checkbox, .radio, .switchRole,
             .slider, .picker, .list, .grid, .scroll, .webview:
            return true
        default:
            return false
        }
    }

    private static func shouldAssignRefToListItem(_ node: SemanticNode) -> Bool {
        if node.states.contains(SemanticStates.clickable)
            || node.states.contains(SemanticStates.clickableInferred) {
            return true
        }
        if node.states.contains(SemanticStates.clickableGuessed) {
            return !hasExecutableDescendant(node)
        }
        return false
    }

    private static func hasExecutableDescendant(_ node: SemanticNode) -> Bool {
        for child in node.children {
            if shouldAssignRef(child, webMode: false) || hasExecutableDescendant(child) {
                return true
            }
        }
        return false
    }

    private final class State {
        let prefix: String
        let webMode: Bool
        var nextRef: Int = 1

        init(prefix: String, webMode: Bool) {
            self.prefix = prefix
            self.webMode = webMode
        }

        func assignNode(_ node: SemanticNode) -> SemanticNode {
            let builder = SemanticNode.builder(node.role)
                .name(node.name)
                .text(node.text)
                .contentDescription(node.contentDescription)
                .resourceId(node.resourceId)
                .className(node.className)
                .bounds(node.bounds)
                .roleDecision(node.roleDecision)
                .webElementIdx(node.webElementIdx)
            for state in node.states {
                builder.addState(state)
            }
            if RefAssigner.shouldAssignRef(node, webMode: webMode) {
                builder.ref(prefix + String(nextRef))
                nextRef += 1
            } else if node.hasRef {
                builder.ref(node.ref)
            }
            for child in node.children {
                builder.addChild(assignNode(child))
            }
            return builder.build()
        }
    }
}

enum SemanticStates {
    static let clickable = "clickable"
    static let clickableInferred = "clickable-inferred"
    static let clickableGuessed = "clickable-guessed"
    static let disabled = "disabled"
    static let checked = "checked"
    static let selected = "selected"
    static let focused = "focused"
    static let scrollable = "scrollable"
    static let password = "password"
}
